include/glow/Support/TensorPool.h

```diff
@@ -0,0 +1,83 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef GLOW_TENSORPOOL_H
+#define GLOW_TENSORPOOL_H
+
+#include "glow/Base/Tensor.h"
+
+#include <iostream>
+#include <mutex>
+#include <unordered_map>
+#include <vector>
+
+namespace glow {
+
+class TensorPool final {
+private:
+  struct TypeHash {
+    size_t operator()(const Type &t) const { return t.hash(); }
+  };
+
+  struct TypeEquals {
+    bool operator()(const Type &a, const Type &b) const { return a.isEqual(b); }
+  };
+  /// A stack of available Tensors per Type.
+  std::unordered_map<Type, std::vector<Tensor>, TypeHash, TypeEquals> pools_;
+
+  /// Mutex around pools_;
+  std::mutex lock_;
+
+  /// Whether or not to allow allocation of new buffers if the pool is empty.
+  const bool preventInlineAllocs_{false};
+
+public:
+  struct Stats {
+    std::atomic<uint64_t> totalTypes{0};
+    std::atomic<uint64_t> totalAllocs{0};
+    std::atomic<uint64_t> inlineAllocs{0};
+    std::atomic<uint64_t> totalGets{0};
+    std::atomic<uint64_t> totalReclaims{0};
+  } stats_;
+
+  TensorPool(bool preventAllocs = false)
+      : preventInlineAllocs_{preventAllocs} {}
+
+  ~TensorPool() { clear(); }
+
+  /// Retrieve a Tensor with type \p ty from the pool - this type must have
+  /// previously been added by initialize. If the pool is empty this will
+  /// allocate a new Tensor unless preventAllocs was provided at construction
+  /// time.
+  Tensor get(TypeRef ty);
```

Comment:
This should probably be "Tensor*". Which is a pointer to some tensor that is managed by the pool.

Reply:
PlaceholderBindings takes Tensors by rvalue ref and takes ownership, so this doesn't work quite how you'd like, I think.